pso/src/main/scala/cilib/pso/PSO.scala

```diff
@@ -54,6 +54,187 @@ object PSO {
     if (b) updatePBest(p) else Step.point(p)
   }
 
+  def clamping[S](particle: Particle[S, Double])(
+    implicit M: HasMemory[S, Double]): Step[Double, Particle[S, Double]] = {
+      val newPos = particle.pos.pos zip particle.pos.boundary map {
+          case t if t._1 < t._2.lowerValue => t._2.lowerValue
+          case t if t._1 > t._2.upperValue => t._2.upperValue
+          case t if t._2.contains(t._1) => t._1
+      }
+
+      for {
+          e <- PSO.evalParticle(Lenses._position.set(Position.apply(newPos, particle.pos.boundary)).apply(particle))
+          u <- PSO.updatePBest(e)
+      } yield u
+  }
+
+  def initToPB[S](particle: Particle[S, Double])(
+    implicit M: HasMemory[S, Double]): Step[Double, Particle[S, Double]] = {
+      val newPos = particle.pos.pos zip particle.pos.boundary zip M._memory.get(particle.state).pos map {
+          case ((a, b), c) => (a, b, c)
+      } map {
+          case t if t._2.contains(t._1) => t._1
+          case t if t._2.doesNotContain(t._1) => t._3
+      }
+
+      for {
+          e <- PSO.evalParticle(Lenses._position.set(Position.apply(newPos, particle.pos.boundary)).apply(particle))
+          u <- PSO.updatePBest(e)
+      } yield u
+  }
+
+  def initToPBAndZeroedVelocity[S](particle: Particle[S, Double])(
+    implicit M: HasMemory[S, Double], V: HasVelocity[S, Double]): Step[Double, Particle[S, Double]] = {
```

Comment:
It seems that composition would be the better strategy here -> could we maybe try see if we can define the zeroing separately and then compose them to get the desired result?